Add replace and treplace to String

ADragonFriend opened the pull request:
I started a little project this weekend to learn C3, and quickly realized that a method to replace substrings would be very useful.

lerno commented:
It is not bad, but I think that perhaps DString.replace could be used instead:
1. Copy the string into a DString
2. Use DString.replace

Could you benchmark the two possibilities?

ADragonFriend commented:
Thanks for the feedback. I hadn't looked into DString very much.

I did some quick benchmarks, and using DString.replace and then DString.copy_str is indeed typically faster with a low number of occurrences of the text to be replaced, but typically a bit slower with a higher number of occurrences.

I also tried using DString.join instead of DString.replace (or String.replace). This was always faster, at least with a lower iteration count. Once I increased the iterations I was measuring to 100, DString.join lost its advantage over the other methods (at least with many occurrences of the text to be replaced).

I also tried replacing spaces in the test string with "-" instead of "--" so the length of the resulting string wouldn't be changed. This hurt DString.join slightly (~1.5% slower) but helped DString.replace (~72% faster).

So I think you're correct that using DString is a much better approach, but I think this simple test doesn't show which of the alternatives I tested is really better right now.

```
import std::io;
import std::time;

fn String replace(String test, String old, String new)
{
	String[] split = test.tsplit(old);
	return string::join(tmem, split, new);
}

fn String dstring_replace(String test, String old, String new)
{
	DString dstr = dstring::temp(test);
	dstr.replace(old, new);
	return dstr.tcopy_str();
}

fn String dstring_join(String test, String old, String new)
{
	String[] split = test.tsplit(old);
	DString dstr = dstring::join(tmem, split, new);
	return dstr.tcopy_str();
}

const int ITERATE = 100;
fn void main()
{
	io::printn("Replace 0 times:");
	NanoDuration diff;
	Clock c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		replace("11 22 33 44 55 66 77 88 99 00", "42", "--");
	}
	diff = c.mark();
	io::printfn("String completed in:\t\t%s", diff);
	
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		dstring_replace("11 22 33 44 55 66 77 88 99 00", "42", "--");
	}
	diff = c.mark();
	io::printfn("DString completed in:\t\t%s", diff);
	
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		dstring_join("11 22 33 44 55 66 77 88 99 00", "42", "--");
	}
	diff = c.mark();
	io::printfn("DString join completed in:\t%s", diff);
	
	//------------------------------
	io::printn("\nReplace 1 time:");
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		replace("11 22 33 44 55 66 77 88 99 00", "55", "--");
	}
	diff = c.mark();
	io::printfn("String completed in:\t\t%s", diff);
	
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		dstring_replace("11 22 33 44 55 66 77 88 99 00", "55", "--");
	}
	diff = c.mark();
	io::printfn("DString completed in:\t\t%s", diff);
	
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		dstring_join("11 22 33 44 55 66 77 88 99 00", "55", "--");
	}
	diff = c.mark();
	io::printfn("DString join completed in:\t%s", diff);
	
	//-------------------------------
	io::printn("\nReplace 9 times:");
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		replace("11 22 33 44 55 66 77 88 99 00", " ", "--");
	}
	diff = c.mark();
	io::printfn("String completed in:\t\t%s", diff);
	
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		dstring_replace("11 22 33 44 55 66 77 88 99 00", " ", "--");
	}
	diff = c.mark();
	io::printfn("DString completed in:\t\t%s", diff);
	
	c = clock::now();
	for(int i = 0; i < ITERATE; i++)
	{
		dstring_join("11 22 33 44 55 66 77 88 99 00", " ", "--");
	}
	diff = c.mark();
	io::printfn("DString join completed in:\t%s", diff);
}

//These are typical results
/*
//ITERATE = 3
Replace 0 times:
String completed in:            15µs268ns
DString completed in:           4µs508ns
DString join completed in:      1µs704ns

Replace 1 time:
String completed in:            5µs39ns
DString completed in:           3µs797ns
DString join completed in:      1µs864ns

Replace 9 times:
String completed in:            5µs320ns
DString completed in:           7µs294ns
DString join completed in:      4µs528ns
*/

/*
//ITERATE = 100
Replace 0 times:
String completed in:            103µs200ns
DString completed in:           118µs858ns
DString join completed in:      69µs558ns

Replace 1 time:
String completed in:            103µs761ns
DString completed in:           116µs955ns
DString join completed in:      77µs513ns

Replace 9 times:
String completed in:            182µs566ns
DString completed in:           137µs422ns
DString join completed in:      162µs198ns
*/

```

lerno commented:
I ran 100000 iterations instead. DString replace were worse, DString.join worked well, esp after fixing some things with the temp pool usage.

lerno commented:
Can you rewrite it using DString under the hood instead?

ADragonFriend commented:
Sure.

Using DString.replace, or using DString.join?

lerno commented:
Join is the one that performs well

lerno commented:
Thank you!
